src/renderers/DomRenderer.js

```diff
@@ -29,7 +29,12 @@ function getStyle(cssRule: HTMLElement|CSSStyleRule, prop: string): string {
  */
 function setStyle(cssRule: HTMLElement|CSSStyleRule, prop: string, value: string): boolean {
   try {
-    cssRule.style.setProperty(prop, value)
+    if (value && value.toLocaleLowerCase().indexOf("!important") >= 0) {
```

Comment:
My problem with this approach is that it will make every update slower, even if it doesn't has an "important". This part of the API is hot, because is used for animations.

Reply:
Second point is that important in general is a bad thing, it is not needed in a clean design, just a hack.

Reply:
Potentially we could support something like this: `color: ['red', '!important']` so an important value needs to be written separately by the user and so we don't have to parse it.

Reply:
First off... wanna say thanks for this awesome library!!!  

While there is definitely some performance hit with the approach, I don't feel that this is or will be the bottleneck of an application when it comes to performance.  Having said that, I don't write graphic intensive applications in Javascript that require thousands of style changes per second.  However, I did setup a test comparison just to show what the impact looks like so that we can talk about it constructively.  The tests can be found [here](https://jsperf.com/jsstolowercasetest/1).  Here's a screenshot as well of a test that I ran:

![image](https://user-images.githubusercontent.com/11198735/32857955-ef5ad1fc-ca06-11e7-9ea7-df9be220695e.png)

Certainly the test called 'No Check' is the fastest (as expected).  However, it's processing over 3M operations per second!!  The test called 'With check and toLowerCase' which is representative of this PR, is ~41% slower, but is still processing over 1.8M operations per second!!  You'd have to be doing a TON of changes to even get close with this being the bottleneck of the application.  The 'As Array' test I think is representative of what you were proposing above... and is certainly faster than the other  approach, but comes with the cost of not being as straight forward to consume.  If we don't do a check and "assume" that '!important' will always be lowercase already then the 'with check' is just as performant as the array approach.

Anyway, long story short, I feel that even though there is a performance hit with this approach, it is the most straight forward to consume (no changes necessary from a consumer standpoint) and is the way that I think everyone would *expect* to happen.  I appreciate the consideration.

Reply:
I am also thinking about other plugins and pipelines. Ideally if we do the conversion from string, we should make it in a way that it is done only once and other plugins can rely on it.

I am thinking that we could support !important conversion to an array as a plugin, so that once it is converted all other plugins and DOM module can rely on arrays.

Reply:
For e.g. we have the same problem in vendor prefixer.

Reply:
What bothers me and I have no solution for yet:

Function values/rules are in JSS high performance, they are capable of being updated at 60fps. Compared to other libraries, they allow only regular one-time updates, everything else will be generating tons of css.

At the same time they are used for one-time updates in jss as well, for e.g. for theming etc.

So we have 2 different usages of this interface, one is very intensive the other one is not so much.

While for regular "theming" usage, such overhead as important parsing or other plugins like verndor prefixer doesn't matter, for high frequency updates it matters.

So I am still not sure how to proceed with them. Evtl. we need an  API change that allows to distinguish those cases and allow optimizing the frequent updates.

Reply:
Also in general I need to decide whether we want to support "!important"  as a mainstream feature. I am not convinced it should be, because there should be no need, but if there is, maybe an optional plugin should do the job?

Reply:
What do you think about supporting array syntax in the core, in setStyle function now. And implement !important   parsing plugin separately?

Reply:
Also, for space separated values core supports double arrays syntax, because single array is for comma separated values, see[

Reply:
I think what you propose with the array syntax makes sense.  I guess I didn't realize that having values as an array was already supported in the core.  And to your point about performance, having a separate plugin makes sense as well since users can choose when to use it and when not to.

Reply:
yep!

Reply:
would you like to send a PR for arrays syntax support?

Reply:
Sure... are you looking for something like:
```
    if (Array.isArray(value) && value[1] === "!important")
      cssRule.style.setProperty(prop, value[0], "important")
    else
      cssRule.style.setProperty(prop, value)
```
My issue with this is that it's not truly supporting array syntax.  Array syntax can be used in many other situations and so it would need to be something more like:
```
    if (Array.isArray(value)) {
      let cssValue = utils.toCssValue(value)
      if (value[value.length] === "!important") {
        cssValue = cssValue.substr(0, cssValue.indexOf("!important"))
        cssRule.style.setProperty(prop, cssValue, "important")
      }
      else
        cssRule.style.setProperty(prop, cssValue)
    }
    else
      cssRule.style.setProperty(prop, value)
```
This seems a lot more verbose and less performant but would handle all the different scenarios where an array might  be used as a value.  Although it would really only be less performant for values that are an array and also even less performant for those that use "!important".  I think it's an acceptable trade-off.  Thoughts?

Reply:
The first version, but checking for the last element in the array instead.

Reply:
actually wait, you are right, there is more to this

Reply:
Ok, we def. need to use toCssValue here. I am thinking of 2 different implementations which can have different perf characteristics:

1. always return an array at the end out of toCssValue and relay on Array#toString.
2. create a class CSSValue and return it. It's toString would to what toCssValue does now + you can check for important flag on the instance

Reply:
On the other hand, you don't need indexOf call, you already know when it is at the end, you can just do substr.

Reply:
So something like:
```
    if (Array.isArray(value)) {
      let cssValue = utils.toCssValue(value)
      if (value[value.length] === "!important") {
        cssValue = cssValue.substr(0, cssValue.length - 10)
        cssRule.style.setProperty(prop, cssValue, "important")
      }
      else
        cssRule.style.setProperty(prop, cssValue)
    }
    else
      cssRule.style.setProperty(prop, value)
```

Reply:
yeah, looks correct and introduces only overhead if there is array at first and then !important at second. Also it supports now a full array syntax, which was basically broken otherwise.